### PR TITLE
Draft demo

### PR DESCRIPTION
**Pull request by jemmyshin**

### CODE REVIEW

File: client/clip_client/client.py [hunk 1]
```diff
@@ -61,6 +62,7 @@ def __init__(self, server: str, credential: dict = {}, **kwargs):
             raise ValueError(f'{server} is not a valid scheme')
 
         self._authorization = credential.get('Authorization', None)
+        self._batch_size = 128
```
Comment (numb3r3) [inline client/clip_client/client.py:65]: please remove the unused codes

File: client/clip_client/client.py [hunk 2]
```diff
@@ -284,9 +289,14 @@ async def aencode(self, content, **kwargs):
         )
 
         results = DocumentArray()
-        async for da in self._async_client.post(
-            **self._get_post_payload(content, kwargs)
-        ):
+
+        parameters = kwargs.get('parameters', {})
```
Comment (numb3r3) [inline client/clip_client/client.py:293]: I prefer to use `kwargs.pop()` instead

File: client/clip_client/client.py [hunk 3]
```diff
@@ -284,9 +289,14 @@ async def aencode(self, content, **kwargs):
...
+        model_name = parameters.get('model', '')
+        payload = self._get_post_parameters(content, kwargs)
+        payload.update(on=f'/encode/{model_name}'.rstrip('/'))
+        payload.update(inputs=self._iter_doc(content))
+
```
Comment (numb3r3) [inline client/clip_client/client.py:298]: The `update` is not needed any more. You can directly pass the parameters `on=...` and `inputs` in `post`

File: client/clip_client/client.py [hunk 4]
```diff
@@ -428,3 +439,54 @@ async def arank(self, docs: Iterable['Document'], **kwargs) -> 'DocumentArray':
...
+        by default.
+        :param content: docs to be indexed.
+        """
+        self._prepare_streaming(
+            not kwargs.get('show_progress'),
+            total=len(content) if hasattr(content, '__len__') else None,
+        )
+
+        payload = self._get_post_parameters(content, kwargs)
+        payload.update(on='/index')
```
Comment (numb3r3) [inline client/clip_client/client.py:456]: same

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: client/clip_client/client.py [hunk 5]
```diff
@@ -112,20 +111,20 @@ def encode(self, content, **kwargs):
...
-        payload.update(on=f'/encode/{model_name}'.rstrip('/'))
-        payload.update(inputs=self._iter_doc(content))
 
         with self._pbar:
             self._client.post(
                 **payload,
+                on=f'/encode/{model_name}'.rstrip('/'),
+                inputs=self._iter_doc(content),
                 on_done=partial(self._gather_result, results=results),
             )
```
Comment (numb3r3) [inline client/clip_client/client.py:129]:
```suggestion
            self._client.post(
                on=f'/encode/{model_name}'.rstrip('/'),
                inputs=self._iter_doc(content),
                on_done=partial(self._gather_result, results=results),
                **payload,
            )
```

File: client/clip_client/client.py [hunk 6]
```diff
@@ -389,22 +390,22 @@ def rank(self, docs: Iterable['Document'], **kwargs) -> 'DocumentArray':
...
         payload = self._get_post_parameters(docs, kwargs)
-        payload.update(on=f'/rank/{model_name}'.rstrip('/'))
-        payload.update(
-            inputs=self._iter_rank_docs(docs, _source=kwargs.get('source', 'matches'))
-        )
 
         with self._pbar:
             self._client.post(
                 **payload,
+                on=f'/rank/{model_name}'.rstrip('/'),
```
Comment (numb3r3) [inline client/clip_client/client.py:405]: same

**Review by numb3r3 — CHANGES_REQUESTED** [review]

**Comment by numb3r3** [comment]
@jemmyshin Pleas rebase your PR to fix the conflict.

File: client/clip_client/client.py [hunk 7]
```diff
@@ -110,14 +111,21 @@ def encode(self, content, **kwargs):
...
         self._prepare_streaming(
-            not kwargs.get('show_progress'),
+            not kwargs.pop('show_progress'),
             total=len(content) if hasattr(content, '__len__') else None,
         )
         results = DocumentArray()
+
+        parameters = kwargs.pop('parameters', {})
+        model_name = parameters.get('model', '')
+        payload = self._get_post_parameters(content, kwargs)
```
Comment (numb3r3) [inline client/clip_client/client.py:121]: rename it as `params`?

File: client/clip_client/client.py [hunk 8]
```diff
@@ -436,3 +448,47 @@ async def arank(self, docs: Iterable['Document'], **kwargs) -> 'DocumentArray':
                 )
 
         return results
+
+    def index(self, content: Iterable['Document'], **kwargs):
+        """Index the embeddings created by server CLIP model.
+        Given the document with embeddings, this function create an indexer which index
```
Comment (numb3r3) [inline client/clip_client/client.py:454]: The docstr is not corrct. `index` and `query` functions both accept raw text/image docs without embeddings.

Comment (numb3r3) [reply client/clip_client/client.py:454]: What's more, we also implment async version, `aindex`, `aquery`

Comment (numb3r3) [reply client/clip_client/client.py:454]: The `index` should have the same function signature as `encode`

File: client/clip_client/client.py [hunk 9]
```diff
@@ -436,3 +448,47 @@ async def arank(self, docs: Iterable['Document'], **kwargs) -> 'DocumentArray':
...
+            not kwargs.pop('show_progress'),
+            total=len(content) if hasattr(content, '__len__') else None,
+        )
+
+        payload = self._get_post_parameters(content, kwargs)
+
+        with self._pbar:
+            self._client.post(on='/index', inputs=self._iter_doc(content), **payload)
+
+    def search(self, content: List[str], **kwargs) -> DocumentArray:
```
Comment (numb3r3) [inline client/clip_client/client.py:469]: `search` should also accept either `str` and `Document`

**Review by numb3r3 — CHANGES_REQUESTED** [review]
We should take care of the user experience. Let's polish the client API to offer an easy-to-use client regarding to user's perspective.

File: client/clip_client/client.py [hunk 10]
```diff
@@ -110,14 +111,21 @@ def encode(self, content, **kwargs):
             )
 
         self._prepare_streaming(
-            not kwargs.get('show_progress'),
+            not kwargs.pop('show_progress'),
             total=len(content) if hasattr(content, '__len__') else None,
         )
         results = DocumentArray()
+
+        parameters = kwargs.pop('parameters', {})
```
Comment (numb3r3) [inline client/clip_client/client.py:119]: alright, `pop` operator would raise an exception when the given `key` does not exist.

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: client/clip_client/client.py [hunk 11]
```diff
@@ -89,35 +91,43 @@ def encode(
     def encode(
         self,
         content: Union['DocumentArray', Iterable['Document']],
+        parameters: Optional[dict] = {},
```
Comment (numb3r3) [inline client/clip_client/client.py:94]: same

File: client/clip_client/client.py [hunk 12]
```diff
@@ -89,35 +91,43 @@ def encode(
...
         :param show_progress: if set, show a progress bar
-        :param parameters: the parameters for the encoding, you can specify the model to use when you have multiple models
+        :param parameters: the parameters for the encoding. Now we support:
+           - model: you can specify the model to use when you have multiple models
         :return: the embedding in a numpy ndarray with shape ``[N, D]``. ``N`` is in the same length of ``content``
         """
         ...
 
-    def encode(self, content, **kwargs):
+    def encode(self, content, parameters, **kwargs):
```
Comment (numb3r3) [inline client/clip_client/client.py:109]: it should be optional

File: client/clip_client/client.py [hunk 13]
```diff
@@ -89,35 +91,43 @@ def encode(
...
         if isinstance(content, str):
             raise TypeError(
                 f'content must be an Iterable of [str, Document], try `.encode(["{content}"])` instead'
             )
 
         self._prepare_streaming(
-            not kwargs.get('show_progress'),
+            not kwargs.pop('show_progress', False),
```
Comment (numb3r3) [inline client/clip_client/client.py:116]: `pop` would raise an exception if not existing. Hence, we should use get here.

File: client/clip_client/client.py [hunk 14]
```diff
@@ -89,35 +91,43 @@ def encode(
...
+        params = self._get_post_parameters(content, kwargs)
+
         with self._pbar:
             self._client.post(
-                **self._get_post_payload(content, kwargs),
+                on=f'/encode/{model_name}'.rstrip('/'),
+                inputs=self._iter_doc(content),
+                parameters=parameters,
                 on_done=partial(self._gather_result, results=results),
+                **params,
```
Comment (numb3r3) [inline client/clip_client/client.py:130]: It seems the `parameters` and `params` is very confusing.

Comment (numb3r3) [reply client/clip_client/client.py:130]: should we use `post_args` instead of `params`?

File: client/clip_client/client.py [hunk 15]
```diff
@@ -259,6 +265,7 @@ def make_table(_title, _time, _percent):
     async def aencode(
         self,
         content: Iterator[str],
+        parameters: Optional[dict] = {},
```
Comment (numb3r3) [inline client/clip_client/client.py:268]: change the function signature

File: client/clip_client/client.py [hunk 16]
```diff
@@ -269,34 +276,45 @@ async def aencode(
...
     ) -> 'DocumentArray':
         ...
 
-    async def aencode(self, content, **kwargs):
+    async def aencode(self, content, parameters, **kwargs):
         from rich import filesize
 
         self._prepare_streaming(
-            not kwargs.get('show_progress'),
+            not kwargs.pop('show_progress', False),
```
Comment (numb3r3) [inline client/clip_client/client.py:290]: double-check whether `pop` will work as we expected

File: client/clip_client/client.py [hunk 17]
```diff
@@ -369,62 +387,312 @@ def _iter_rank_docs(
...
+        self,
+        content: Iterable[str],
+        parameters: Optional[dict] = {},
+        *,
+        batch_size: Optional[int] = None,
+        show_progress: bool = False,
+    ) -> 'DocumentArray':
+        """Search for top k results for given query string or ``Document``.
+        If the input is a string, will use this string as query. If the input is a
+        ``Document``, will use the ``text`` field as query.
```
Comment (numb3r3) [inline client/clip_client/client.py:579]: No, `Document` can also be an image doc. The docstr is misleading

File: client/clip_client/client.py [hunk 18]
```diff
@@ -369,62 +387,312 @@ def _iter_rank_docs(
...
+        :param parameters: parameters passed to search function. Now we support:
+            - limit: int, return top limit results. Default is 10.
+            - filter: dict, apply filter when querying. Default is None.
+            - include_metadata: bool, whether return the document metadata in response. Default is True.
+        :param batch_size: the number of elements in each request when sending ``content``.
+        :param show_progress: if set, show a progress bar.
+        :param parameters: parameters passed to search function. Now we support:
+            - limit: int, return top limit results. Default is 10.
+            - filter: dict, apply filter when querying. Default is None.
+            - include_metadata: bool, whether return the document metadata in response. Default is True.
```
Comment (numb3r3) [inline client/clip_client/client.py:590]: the repeated item.

File: client/clip_client/client.py [hunk 19]
```diff
@@ -369,62 +387,312 @@ def _iter_rank_docs(
...
+    def search(
+        self,
+        content: Union['DocumentArray', Iterable['Document']],
+        parameters: Optional[dict] = {},
+        *,
+        batch_size: Optional[int] = None,
+        show_progress: bool = False,
+    ) -> 'DocumentArray':
+        """Search for top k results for given query string or ``Document``.
+        If the input is a string, will use this string as query. If the input is a
```
Comment (numb3r3) [inline client/clip_client/client.py:605]: same

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: client/clip_client/client.py [hunk 20]
```diff
@@ -114,10 +117,17 @@ def encode(self, content, **kwargs):
             total=len(content) if hasattr(content, '__len__') else None,
         )
         results = DocumentArray()
+
+        model_name = parameters.get('model', '')
```
Comment (numb3r3) [inline client/clip_client/client.py:121]: raise error when `parameters=None`

**Review by numb3r3 — CHANGES_REQUESTED** [review]

**Comment by numb3r3** [comment]
@jemmyshin please rebase your PR to fix conflicts

File: client/clip_client/client.py [hunk 21]
```diff
@@ -284,9 +292,16 @@ async def aencode(self, content, **kwargs):
         )
 
         results = DocumentArray()
+
+        model_name = parameters.get('model', '')
```
Comment (numb3r3) [inline client/clip_client/client.py:296]: `parameters` can be None, and will raise error here. Please fix it

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: server/clip_server/demo-flow.yml [hunk 22]
```diff
@@ -0,0 +1,22 @@
...
+      metas:
+        py_modules:
+          - clip_server.executors.clip_torch
+    replicas: 1
+  - name: indexer
+    uses:
+      jtype: AnnLiteIndexer
+      with:
+        dim: 512
+        data_path: /Users/jemfu/Desktop/laion400m/workspace
```
Comment (numb3r3) [inline server/clip_server/demo-flow.yml:18]: don't use absolute real path

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: client/clip_client/client.py [hunk 23]
```diff
@@ -369,62 +387,308 @@ def _iter_rank_docs(
...
+        self._prepare_streaming(
+            not kwargs.get('show_progress'),
+            total=len(content) if hasattr(content, '__len__') else None,
+        )
+        results = DocumentArray()
+        post_args = self._get_post_args(content, kwargs)
+
+        with self._pbar:
+            self._client.post(
+                on='/index',
```
Comment (ZiniuYu) [inline client/clip_client/client.py:513]: will we also support selection from multiple indexers?

**Comment by numb3r3** [comment]
@jemmyshin This PR has been merged to https://github.com/jina-ai/clip-as-service/pull/816. I will close this PR first. let's work on the new PR.

**Pull request closed without merging**